### PR TITLE
Mutation Testing

### PR DESCRIPTION
**Pull request by sigmade**
## Summary

Added an article about Mutation Testing

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/testing/mutation-testing.md](https://github.com/dotnet/docs/blob/5f759b202e041f72f6935d19f97c1ea3ea845bdf/docs/core/testing/mutation-testing.md) | [Mutation testing](https://review.learn.microsoft.com/en-us/dotnet/core/testing/mutation-testing?branch=pr-en-us-44615) |
| [docs/navigate/devops-testing/toc.yml](https://github.com/dotnet/docs/blob/5f759b202e041f72f6935d19f97c1ea3ea845bdf/docs/navigate/devops-testing/toc.yml) | [docs/navigate/devops-testing/toc](https://review.learn.microsoft.com/en-us/dotnet/navigate/devops-testing/toc?branch=pr-en-us-44615) |

### CODE REVIEW

**Comment by IEvangelist** [comment]
> @IEvangelist can you please check?

I'm reviewing this, thank you for your patience.

**Review by IEvangelist — COMMENTED** [review]
Let's get these suggestions in and I'll review it again.

File: docs/core/testing/mutation-testing.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+{
+    var exception = Assert.Throws<ArgumentException>(()
+        => PriceCalculator.CalculatePrice(0, 10));
+    Assert.Equal("Price must be greater than zero.", exception.Message);
+}
+```
+
+:::image type="content" source="media/stryker-final-report.png" lightbox="media/stryker-final-report.png" alt-text="Stryker final report":::
+
+Mutation testing helps to find weak points in tests and makes them more reliable. It forces you to check not only the 'happy path', but also complex boundary cases, reducing the likelihood of bugs in production.
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:154]:
```suggestion
Mutation testing helps to find opportunities to improve tests that make them more reliable. It forces you to check not only the 'happy path', but also complex boundary cases, reducing the likelihood of bugs in production.
```

File: docs/core/testing/mutation-testing.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+```
+
+First, install **Stryker.NET**.
+To do this, you need to execute the command:
+
+```dotnetcli
+dotnet tool install -g dotnet-stryker
+```
+
+Let's run Stryker using the command from the directory where the unit tests are located:
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:80]:
```suggestion
To run `stryker`, invoke it from the command line in the directory where the unit test project is located:
```

File: docs/core/testing/mutation-testing.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+```dotnetcli
+dotnet stryker
+```
+
+After the testing is completed, a report will be displayed in the console - how many mutants were killed and how many survived.
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:86]:
```suggestion
After the tests have run, a report is displayed in the console.
```

File: docs/core/testing/mutation-testing.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+[Fact]
+public void InvalidPrice_ShouldThrowException()
+{
+    var calculator = new PriceCalculator();
+
+    Assert.Throws<ArgumentException>(() => calculator.CalculatePrice(-10, 10));
+}
+```
+
+First, install **Stryker.NET**.
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:73]:
```suggestion

The preceding code highlights two projects, one for the service that acts as a `PriceCalculator` and the other is the test project.

## Install the global tool

First, install **Stryker.NET**.
```

File: docs/core/testing/mutation-testing.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+After the tests have run, a report is displayed in the console.
+
+:::image type="content" source="media/stryker-console-report.png" lightbox="media/stryker-console-report.png" alt-text="Stryker console report":::
+
+**Stryker.NET** saves a detailed HTML report in the StrykerOutput directory.
+
+:::image type="content" source="media/stryker-first-report.png" lightbox="media/stryker-first-report.png" alt-text="Stryker first report":::
+
+Now, consider what mutants are and what 'survived' and 'killed' mean. A mutant is a small change in your code that Stryker makes on purpose. The idea is simple: if your tests are good, they should catch the change and fail. If they still pass, your tests might not be strong enough.
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:94]:
```suggestion
Now, consider what mutants are and what 'survived' and 'killed' mean. A mutant is a small change in your code that Stryker makes on purpose. The idea is simple: if your tests are good, they should catch the change and fail. If they still pass, your tests might not be strong enough.
```

File: docs/core/testing/mutation-testing.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+| Type | Description |
+|--|--|
+| Equivalent | The equivalent operator is used to replace an operator with its equivalent. For example, `x < y` becomes `x <= y`. |
+| Arithmetic | The arithmetic operator is used to replace an arithmetic operator with its equivalent. For example, `x + y` becomes `x - y`. |
+| String | The string operator is used to replace a string with its equivalent. For example, `"text"` becomes `""`. |
+| Logical | The logical operator is used to replace a logical operator with its equivalent. For example, `x && y` becomes `x \|\| y`. |
+
+For additional mutation types, see the [Stryker.NET: Mutations](https://stryker-mutator.io/docs/stryker-net/mutations) documentation.
+
+If, after changing your code, the unit tests pass successfully, then they aren't sufficiently robust, and the mutant survived.
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:109]:
```suggestion
## Incremental improvement

If, after changing your code, the unit tests pass successfully, then they aren't sufficiently robust, and the mutant survived.
```

**Comment by IEvangelist** [comment]
@sigmade - there's a merge conflict with the recent changes to include Copilot generation of tests - could you please address that?

**Comment by sigmade** [comment]
> @sigmade - there's a merge conflict with the recent changes to include Copilot generation of tests - could you please address that?

Done

File: docs/core/testing/mutation-testing.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+[Fact]
+public void InvalidPrice_ShouldThrowException()
+{
+    var calculator = new PriceCalculator();
+
+    Assert.Throws<ArgumentException>(() => calculator.CalculatePrice(-10, 10));
+}
+```
+
```
Comment (IEvangelist) [inline docs/core/testing/mutation-testing.md:72]:
```suggestion
```

**Review by IEvangelist — APPROVED** [review]
